packages/fluentui/react-northstar/src/components/Attachment/Attachment.tsx

```diff
@@ -61,104 +70,161 @@ export interface AttachmentProps extends UIComponentProps, ChildrenComponentProp
-      e.preventDefault();
-      return;
-    }
-
-    _.invoke(props, 'onClick', e, props);
-  };
-
-  const element = (
-    <ElementType {...getA11Props('root', { className: classes.root, onClick: handleClick, ...unhandledProps })}>
-      {AttachmentIcon.create(icon)}
-
-      {(header || description) &&
-        AttachmentBody.create(body, {
-          overrideProps: {
-            content: (
-              <>
-                {AttachmentHeader.create(header)}
-                {AttachmentDescription.create(description)}
-              </>
-            ),
-          },
+    } = props;
+
+    const getA11Props = useAccessibility(accessibility, {
+      debugName: composeOptions.displayName,
+      actionHandlers: {
+        performClick: e => {
+          if (e.currentTarget === e.target) {
+            e.stopPropagation();
+            handleClick(e);
+          }
+        },
+      },
+      rtl: context.rtl,
+    });
+
+    const { classes } = useStyles<AttachmentStylesProps>(composeOptions.displayName, {
+      className: attachmentClassName,
+      mapPropsToStyles: () => ({
+        actionable: actionable || !!onClick,
+        disabled,
+      }),
+      mapPropsToInlineStyles: () => ({
+        className,
+        design,
+        styles,
+        variables,
+      }),
+      rtl: context.rtl,
+      composeOptions,
+      unstable_props: props,
+    });
+
+    const slotProps = composeOptions.resolveSlotProps<AttachmentProps>(props);
+
+    const ElementType = getElementType(props);
+    const unhandledProps = useUnhandledProps(composeOptions.handledProps, props);
+
+    const handleClick = (e: React.KeyboardEvent | React.MouseEvent) => {
+      if (disabled) {
+        e.preventDefault();
+        return;
+      }
+
+      _.invoke(props, 'onClick', e, props);
+    };
+
+    const element = (
+      <ElementType {...getA11Props('root', { className: classes.root, onClick: handleClick, ref, ...unhandledProps })}>
+        {createShorthand(composeOptions.slots.icon, icon, {
+          defaultProps: () => slotProps.icon || {},
+          overrideProps: predefinedProps => ({
+            variables: mergeComponentVariables(variables, predefinedProps.variables),
+          }),
         })}
 
-      {AttachmentAction.create(action)}
-      {!_.isNil(progress) && <div className="ui-attachment__progress" style={{ width: `${progress}%` }} />}
-    </ElementType>
-  );
-  setEnd();
-
-  return element;
+        {(header || description) &&
+          createShorthand(composeOptions.slots.body, body, {
+            defaultProps: () => slotProps.body || {},
+            overrideProps: predefinedProps => ({
+              content: (
+                <>
+                  {createShorthand(composeOptions.slots.header, header, {
+                    defaultProps: () => slotProps.header || {},
+                    overrideProps: predefinedProps => ({
+                      variables: mergeComponentVariables(variables, predefinedProps.variables),
```

Comment:
Do we have a short circuit in `mergeComponentVariables` (`if (!a && !b) return undefined`)? As I remember it will always create a function and break caching

Reply:
Let me check that, good call

Reply:
Yeah, it was breaking the caching, should be fixed now